src/fft.jl

```diff
@@ -1,63 +1,121 @@
+    # order to just have one G vector for 1D or 2D systems)
+    Glims = [0, 0, 0]
+    for i = 1:3
+        if Glims_[i] != 0
+            Glims[i] = ceil(Int, Glims_[i] .- tol)
+        end
+    end
+    Glims
+end
+
+_smallprimes() = [2, 3, 5, 7]
+
 @doc raw"""
 Determine the minimal grid size for the cubic basis set to be able to
 represent product of orbitals (with the default `supersampling=2`).
 
 Optionally optimize the grid afterwards for the FFT procedure by
 ensuring factorization into small primes.
 
-The function will determine the smallest cube containing the wave vectors
+The function will determine the smallest parallelepiped containing the wave vectors
  ``|G|^2/2 \leq E_\text{cut} ⋅ \text{supersampling}^2``.
 For an exact representation of the density resulting from wave functions
 represented in the spherical basis sets, `supersampling` should be at least `2`.
 """
-function determine_grid_size(lattice::AbstractMatrix{T}, Ecut; supersampling=2, tol=1e-8,
-                             ensure_smallprimes=true) where T
-    cutoff_Gsq = 2 * supersampling^2 * Ecut
-    Gmax = [norm(lattice[:, i]) / 2T(π) * sqrt(cutoff_Gsq) for i in 1:3]
+function determine_fft_size(lattice::AbstractMatrix{T}, Ecut; supersampling=2, tol=1e-8,
+                            ensure_smallprimes=true) where T
+    Gmax = supersampling * sqrt(2 * Ecut)
+    Glims = bounding_rectangle(lattice, Gmax; tol=tol)
 
-    # Round up, unless exactly zero (in which case keep it zero in
-    # order to just have one G vector for 1D or 2D systems)
-    for i = 1:3
-        if Gmax[i] != 0
-            Gmax[i] = ceil.(Int, Gmax[i] .- tol)
+    fft_size = Vec3(2 .* Glims .+ 1)
+    # Optimize FFT grid size: Make sure the number factorises in small primes only
+    if ensure_smallprimes
+        fft_size = nextprod.(Ref(_smallprimes()), fft_size)
+    end
+    fft_size
+end
+function determine_fft_size(model::Model, Ecut; kwargs...)
+    determine_fft_size(model.lattice, Ecut; kwargs...)
+end
+function lattice_diameter(latt)
+    diam = zero(eltype(latt))
+    # brute force search
+    for i in (-1, 0, 1)
+        for j in (-1, 0, 1)
+            for k in (-1, 0, 1)
+                diam = max(diam, norm(latt*[i; j; k]))
+            end
         end
     end
+    diam
+end
 
+## This uses a more precise and slower algorithm than the one above,
+## simply enumerating all G vectors and seeing where their difference
+## is. It needs the kpoints to do so.
+function determine_fft_size_precise(lattice::AbstractMatrix{T}, Ecut, kpoints;
+                                    supersampling=2, ensure_smallprimes=true) where T
+    recip_lattice = 2T(π)*pinv(lattice')  # pinv in case one of the dimension is trivial
+    recip_diameter = lattice_diameter(recip_lattice)
+    Glims = [0, 0, 0]
+    # get the bounding rectangle the naive way
+    for kpt in kpoints
+        for G in G_vectors(kpt)
+            if norm(recip_lattice * (G+kpt.coordinate)) ≤ sqrt(2Ecut) - recip_diameter
```

Comment:
space around +

Reply:
This is typically the kind of places where I like no spaces, to indicate priority (even with parentheses). But OK